storage/allocator.go

```diff
@@ -13,75 +13,291 @@
-		// TODO(bdarnell): use a real random seed.
-		rand: *rand.New(rand.NewSource(0)),
+// newAllocator creates a new allocator for the store specified by
+// storeID.
+func newAllocator(g *gossip.Gossip) *allocator {
+	a := &allocator{
+		gossip:  g,
+		randGen: rand.New(rand.NewSource(rand.Int63())),
+	}
+	// Callback triggers on any capacity gossip updates.
+	if a.gossip != nil {
+		capacityRegex := gossip.MakePrefixPattern(gossip.KeyCapacityPrefix)
+		a.gossip.RegisterCallback(capacityRegex, a.capacityGossipUpdate)
 	}
+	return a
 }
 
-// allocate returns a suitable store based on the supplied
-// attributes list. If none are available / suitable, returns an
-// error. It uses the allocator's StoreFinder to select the set of
-// available stores matching attributes for missing replicas and picks
-// using randomly weighted selection based on available capacities.
-func (a *allocator) allocate(required proto.Attributes, existingReplicas []proto.Replica) (
-	*proto.StoreDescriptor, error) {
+// getUsedNodes returns a set of node IDs which are already being used
+// to store replicas.
+func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
 	// Get a set of current nodes -- we never want to allocate on an existing node.
-	usedNodes := make(map[proto.NodeID]struct{})
-	for _, replica := range existingReplicas {
+	usedNodes := map[proto.NodeID]struct{}{}
+	for _, replica := range existing {
 		usedNodes[replica.NodeID] = struct{}{}
 	}
+	return usedNodes
+}
+
+// storeDescFromGossip retrieves a StoreDescriptor from the specified
+// capacity gossip key. Returns an error if the gossip doesn't exist
+// or is not a StoreDescriptor.
+func storeDescFromGossip(key string, g *gossip.Gossip) (*proto.StoreDescriptor, error) {
+	info, err := g.GetInfo(key)
 
-	stores, err := a.storeFinder(required)
 	if err != nil {
 		return nil, err
 	}
+	storeDesc, ok := info.(proto.StoreDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("gossiped info is not a StoreDescriptor: %+v", info)
+	}
+	return &storeDesc, nil
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, contentsChanged bool) {
+	a.Lock()
+	defer a.Unlock()
+
+	// Clear the cached store lists on new gossip.
+	a.storeLists = nil
+	if a.capacityKeys == nil {
+		a.capacityKeys = map[string]struct{}{}
+	}
+	a.capacityKeys[key] = struct{}{}
+}
 
-	// Randomly pick a node weighted by capacity.
-	var candidates []*proto.StoreDescriptor
-	var capacityTotal float64
+// Allocate returns a suitable store for a new allocation with the
+// required attributes. Nodes already accommodating existing replicas
+// are ruled out as targets.
+func (a *allocator) Allocate(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
```

Comment:
Re: absolute space...one step at a time. I actually doubt absolute space will be something we ever get around to worrying about. The next item will be iops load rebalancing.

The explanation for 3 random stores is in the struct doc for `allocator`. It's an attempt to be very random with placement, but provide a bias towards balancing.

Thing is, based on how our clusters grow by splitting in situ, the straight up allocation is relatively rare and comes into play only when there's a machine or disk death. Otherwise, we're always rebalancing.